File: Template.Actionable.Domain.DAL/src/EventSourcing.fs
module Template.Actionable.Domain.DAL.Template.ActionableEventStore
open Template.Actionable.Data.Models
open Common.FSharp.Envelopes
open Newtonsoft.Json
open Microsoft.EntityFrameworkCore


type ActionableDbContext with 
    member this.GetAggregateEvents<'a,'b when 'b :> EnvelopeEntityBase and 'b: not struct>
        (dbset:ActionableDbContext->DbSet<'b>)
        (StreamId.Id (aggregateId):StreamId)
        :seq<Envelope<'a>>= 
        query {
            for event in this |> dbset do
            where (event.StreamId = aggregateId)
            select event
        } |> Seq.map (fun event ->
            {
                Id = event.Id
                UserId = UserId.box event.UserId
                StreamId = StreamId.box aggregateId
                TransactionId = TransId.box event.TransactionId
                Version = Version.box (event.Version)
                Created = event.TimeStamp
                Item = (JsonConvert.DeserializeObject<'a> event.Event)
            })

open Template.Actionable.Domain.WidgetManagement
type WidgetManagementEventStore () =
    interface IEventStore<WidgetManagementEvent> with
        member this.GetEvents (streamId:StreamId) =
            use context = new  ActionableDbContext ()
            streamId
            |> context.GetAggregateEvents (fun i -> i.WidgetEvents) 
            |> Seq.toList 
            |> List.sortBy(fun x -> x.Version)
        member this.AppendEvent (envelope:Envelope<WidgetManagementEvent>) =
            try
                use context = new ActionableDbContext ()
                context.WidgetEvents.Add (
                    WidgetEventEnvelopeEntity (  Id = envelope.Id,
                                            StreamId = StreamId.unbox envelope.StreamId,
                                            UserId = UserId.unbox envelope.UserId,
                                            TransactionId = TransId.unbox envelope.TransactionId,
                                            Version = Version.unbox envelope.Version,
                                            TimeStamp = envelope.Created,
                                            Event = JsonConvert.SerializeObject(envelope.Item)
                                            )) |> ignore         
                context.SaveChanges () |> ignore
                
            with
                | ex -> System.Diagnostics.Debugger.Break () 

File: Template.Actionable.Data/Models/EventSourcing.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Template.Actionable.Data.Models
{
    public class EventStream
    {
        [Column(Order = 0), Key]
        public virtual long Id { get; set; }
        public virtual System.DateTimeOffset Created { get; set; }
    }
    public abstract class EnvelopeEntityBase
    {
        public virtual System.Guid StreamId { get; set; }
        public virtual EventStream Stream { get; set; }
        public virtual System.Guid UserId { get; set; }
        public virtual System.Guid Id { get; set; }

        public virtual System.Guid TransactionId { get; set; }
        public virtual string DeviceId { get; set; }
        public virtual short Version { get; set; }
        public virtual System.DateTimeOffset TimeStamp { get; set; }
        public virtual string Event { get; set; }
    }
    public class WidgetEventEnvelopeEntity : EnvelopeEntityBase { }
}
